autoPyTorch/ensemble/ensemble_builder.py

```diff
@@ -0,0 +1,1432 @@
+                and max_models_on_disc. Might return less
+            seed: int
+                random seed
+            precision: [16,32,64,128]
+                precision of floats to read the predictions
+            memory_limit: Optional[int]
+                memory limit in mb. If ``None``, no memory limit is enforced.
+            read_at_most: int
+                read at most n new prediction files in each iteration
+            logger_name: str
+                Name of the logger where we are gonna write information
+        """
+
+        super(EnsembleBuilder, self).__init__()
+
+        self.backend = backend  # communication with filesystem
+        self.dataset_name = dataset_name
+        self.task_type = task_type
+        self.output_type = output_type
+        self.metrics = metrics
+        self.opt_metric = opt_metric
+        self.ensemble_size = ensemble_size
+        self.performance_range_threshold = performance_range_threshold
+
+        if isinstance(ensemble_nbest, numbers.Integral) and ensemble_nbest < 1:
+            raise ValueError("Integer ensemble_nbest has to be larger 1: %s" %
+                             ensemble_nbest)
+        elif not isinstance(ensemble_nbest, numbers.Integral):
+            if ensemble_nbest < 0 or ensemble_nbest > 1:
+                raise ValueError(
+                    "Float ensemble_nbest best has to be >= 0 and <= 1: %s" %
+                    ensemble_nbest)
+
+        self.ensemble_nbest = ensemble_nbest
+
+        # max_models_on_disc can be a float, in such case we need to
+        # remember the user specified Megabytes and translate this to
+        # max number of ensemble models. max_resident_models keeps the
+        # maximum number of models in disc
+        if max_models_on_disc is not None and max_models_on_disc < 0:
+            raise ValueError(
+                "max_models_on_disc has to be a positive number or None"
+            )
+        self.max_models_on_disc = max_models_on_disc
+        self.max_resident_models = None  # type: Optional[int]
+
+        self.seed = seed
+        self.precision = precision
+        self.memory_limit = memory_limit
+        self.read_at_most = read_at_most
+        self.random_state = check_random_state(random_state)
+
+        # Setup the logger
+        self.logger_name = logger_name
+        self.logger = self._get_ensemble_logger(
+            self.logger_name, self.backend.temporary_directory, False)
+
+        if ensemble_nbest == 1:
+            self.logger.debug("Behaviour depends on int/float: %s, %s (ensemble_nbest, type)" %
+                              (ensemble_nbest, type(ensemble_nbest)))
+
+        self.start_time = 0.0
+        self.model_fn_re = re.compile(MODEL_FN_RE)
+
+        self.last_hash = None  # hash of ensemble training data
+        self.y_true_ensemble = None
+        self.SAVE2DISC = True
+
+        # already read prediction files
```

Comment:
Do we still need this?

Reply:
Yes and I have update the comment as to why:
```
  # already read prediction files
         # We read in back this object to give the ensemble the possibility to have memory
         # Every ensemble task is sent to dask as a function, that cannot take un-picklable
         # objects as attributes. For this reason, we dump to disk the stage of the past
         # ensemble iterations to kick-start the ensembling process

```